Added automatic push distributor selection

ArnyminerZ opened the pull request:
### Purpose

See #1737, but basically, Push is basically stable enough to at least select a distributor by default.

In any case, if one doesn't trust it, the server admin should simply not install support for push on the server, and the distributor will be unused.

### Short description

- Added automatic push distributor selection
  - Added a new optionally injectable class called `PushDistributorDefaults`. It contains a preferred distributor.
  - For OSE, no distributor has preference.
  - For Play Store builds we might want to pre-select FCM for example, or for managed. This can be set with the `DistributorPreferences` implementation.
- Added a new settings entry which will be set to true whenever the user manually disables push (because it will be enabled by default). If this setting is true, no push distributor will be automatically selected.
- On the push distributor selection dialog, the preferred distributor, if any, is marked with "Recommended"
  <details>
    <summary>Screenshot</summary>

    <img width="1080" height="2424" alt="image" src="https://github.com/user-attachments/assets/680922ab-f3b3-41c1-b846-9692fc227302" />
  </details>

We still have to add the individual implementations for each flavor on bitfireAt/davx5.

- Subscriptions are invalidated when the push distributor is changed to make sure they're again registered with the correct endpoint.

### Checklist

- [x] The PR has a proper title, description and label.
- [x] I have [self-reviewed the PR](https://patrickdinh.medium.com/review-your-own-pull-requests-5634cad10b7a).
- [x] I have added documentation to complex functions and functions that can be used by other modules.
- [x] I have added reasonable tests or consciously decided to not add tests.

ArnyminerZ committed:
Added automatic push distributor selection

ArnyminerZ committed:
Merge branch 'main-ose' into 1737-enable-push-by-default

rfc2822 commented on app/src/main/kotlin/at/bitfire/davdroid/push/PushRegistrationManager.kt:
```diff
@@ -102,6 +114,27 @@ class PushRegistrationManager @Inject constructor(
+                    logger.fine("No Push distributor selected and no distributors are available.")
+                }
+            }
+        }
```
I think this code should go to a seperate `PushDistributorManager`, because it's not related to the task of updating the subscriptions.

One class should only do one thing:

- `PushDistributorManager` → manage push distributor
- `PushRegistrationManager` → manage actual subscriptions

The same applies to the existing `PushRegistrationManager.setPushDistributor()`, which should also be moved to `PushDistributorManager`.

---

Then the `PushRegistrationWorker` could

1. `pushDistributorManager.update()` → make sure distributor is correct and set up,
2. `pushRegistrationManager.update()` → update subscriptions with selected push distributor.

rfc2822 reviewed (commented):
I think it will finally be time 😄

As I understand it, the basic idea is to check/update the push distributor whenever the PushRegistrationWorker is run, instead of setting it explicitly from the UI.

This sounds good to me! However some notes.

(Also please fix conflicts)

ArnyminerZ committed:
Merge branch 'main-ose' into 1737-enable-push-by-default

# Conflicts:
#	app/src/ose/kotlin/at/bitfire/davdroid/di/OseFlavorModule.kt

ArnyminerZ committed:
Optimize imports

Signed-off-by: Arnau Mora Gras <arnyminerz@proton.me>

ArnyminerZ committed:
Create `PushDistributorManager`

Signed-off-by: Arnau Mora Gras <arnyminerz@proton.me>

ArnyminerZ committed:
Merge branch 'main-ose' into 1737-enable-push-by-default

ArnyminerZ committed:
Merge branch 'main-ose' into 1737-enable-push-by-default

ArnyminerZ committed:
Merge branch 'main-ose' into 1737-enable-push-by-default

# Conflicts:
#	core/src/androidTest/kotlin/at/bitfire/davdroid/di/TestFlavorModule.kt
#	core/src/main/kotlin/at/bitfire/davdroid/push/PushDistributorManager.kt
#	core/src/main/kotlin/at/bitfire/davdroid/push/PushRegistrationManager.kt
#	core/src/main/kotlin/at/bitfire/davdroid/ui/AppSettingsModel.kt

ArnyminerZ committed:
Add `PushDistributorManager` back

Signed-off-by: Arnau Mora <arnyminerz@proton.me>

ArnyminerZ committed:
Merge remote-tracking branch 'origin/main' into 1737-enable-push-by-default

ArnyminerZ committed:
Move file

Signed-off-by: Arnau Mora <arnyminerz@proton.me>

ArnyminerZ committed:
Merge remote-tracking branch 'refs/remotes/origin/main' into 1737-enable-push-by-default

ArnyminerZ committed:
Update required files

Signed-off-by: Arnau Mora <arnyminerz@proton.me>

ArnyminerZ committed:
Remove remaining log methods

Signed-off-by: Arnau Mora <arnyminerz@proton.me>

rfc2822 committed:
Merge branch 'main' into 1737-enable-push-by-default

rfc2822 committed:
Remove DistributorPreferences and replace with PushDistributorDefaults

rfc2822 committed:
Update push distributor selection and UI improvements

- Simplify push disable setting from `EXPLICIT_PUSH_DISABLE` to `PUSH_DISABLE`
- Improve push distributor list UI with better labels and icons
- Remove unused imports and `AppTheme` wrapper from previews

rfc2822 committed:
Update push registration logic and distributor selection

- Refactor `PushRegistrationManager.update()` to explicitly handle distributor, subscriptions, and worker updates
- Add thread-safety documentation for `update()` and `update(serviceId)`
- Fix push distributor preference comparison in `AppSettingsModel`

rfc2822 committed:
Refactor push distributor logic and improve UI

- Move push disable setting logic to `PushDistributorManager.setPushDistributor()`
- Update comments and docstrings for clarity
- Rename `pushDistributorPreferences` to `pushDistributorDefaults` in DI module

rfc2822 committed:
Optimize push distributor handling and UI performance

- Cache icon bitmap in `AppSettingsScreen` to avoid recomputation
- Add `PushRegistrationManager` to `AppSettingsModel` and update registrations on distributor change
- Improve `PushDistributorManager.getCurrentDistributor()` docs to clarify behavior when distributor is uninstalled

rfc2822 committed:
Update `PushDistributorManager` to clarify distributor selection logic

- Rename `update()` to `updateDistributor()` for consistency
- Improve comments and docstrings in `PushRegistrationManager`
- Add documentation for UnifiedPush callback behavior

rfc2822 committed:
Improve comment clarity for push distributor unregistration behavior

rfc2822 committed:
Invalidate push subscriptions on distributor change

- Add `CollectionDao.invalidatePushSubscriptions()` to clear push subscription expires
- Update `PushDistributorManager` to invalidate subscriptions when distributor changes
- Use application scope in `AppSettingsModel` to prevent interruption during push distributor updates

sunkup reviewed (commented):
So this will do nothing in davx5-ose, because there is no hradcoded/injected preference? 

Also, sounds like this implementation will be superseeded by: https://github.com/bitfireAT/davx5-ose/issues/2091#issuecomment-4108591428 so maybe we should do that instead of merging something unfinished?

I am not quite sure what the requirements for accepting an implementation are. The issue https://github.com/bitfireAT/davx5-ose/issues/1737 does not really look agreed and settled yet.

Also, since we need to work on the flavors across repos, maybe some sub-issues would be good?

ArnyminerZ committed:
Merge branch 'main' into 1737-enable-push-by-default

ArnyminerZ committed:
Merge branch 'main' into 1737-enable-push-by-default

sunkup reviewed (changes requested):
Hmm, currently I can't see any default selection happening in the ose flavor? When installing DAVx5 where ntfy and NextPush are already installed no distributor is chosen by default. 

Maybe I misunderstand something, or test it wrong, since the issue description is still unclear, but how/when should default distributor be selected/changed? When a new distributer is installed, but none existed before? That also does not work in this PR yet. 

It also breaks the detection of some of the available distributors for me. On main branch NextPush is showing up, but using this PR NextPush, does not show up in the list anymore.

This PR:
<img width="574" height="704" alt="Image" src="https://github.com/user-attachments/assets/5c491361-2fa7-4054-aae1-d55a9c8b5e18" />

Main branch:
<img width="574" height="704" alt="Image" src="https://github.com/user-attachments/assets/43a4d2f8-97d8-40ce-b966-e98bc9487148" />

I think it would be good to have a discussion on this together :)

rfc2822 commented:
Close in favor of https://github.com/bitfireAT/davx5-ose/issues/2091
